### PR TITLE
Feat/speed gas

### PR DESCRIPTION
**Pull request by hydraCy**

### CODE REVIEW

File: Maskbook/Scene/Wallet/TokenDetail/View/TransactionCell.swift [hunk 1]
```diff
@@ -75,9 +80,50 @@ class TransactionCell: UITableViewCell {
...
         label.font = FontStyles.BH5
         label.translatesAutoresizingMaskIntoConstraints = false
+        NSLayoutConstraint.activate([
+            label.heightAnchor.constraint(equalToConstant: 24)
+        ])
+
         return label
     }()
     
+    private lazy var refundStackView = HStackView (spacing: 8) {
```
Comment (jk234ert) [inline Maskbook/Scene/Wallet/TokenDetail/View/TransactionCell.swift]: Why it's named `refund`StackView?

File: Maskbook/Scene/Wallet/TokenDetail/View/TransactionCell.swift [hunk 2]
```diff
@@ -209,5 +259,64 @@ class TransactionCell: UITableViewCell {
...
+    @objc
+    func speedUpClicked(_ sender: UIButton) {
+        let result = PendTransactionManager.shared.pendTransactions.value.filter {
+            $0.txHash == self.txHash &&
+            $0.address == maskUserDefaults.defaultAccountAddress &&
+            $0.networkId == maskUserDefaults.network.networkId
+        }
+        
+        if !result.isEmpty {
+            let pendlistModel = result.first
```
Comment (jk234ert) [inline Maskbook/Scene/Wallet/TokenDetail/View/TransactionCell.swift]:
```suggestion
        if let txToSpeedUp = result.first {
```

File: Maskbook/Scene/Wallet/TokenDetail/ViewModel/TokenDetailViewModel.swift [hunk 3]
```diff
@@ -35,19 +35,27 @@ class TokenDetailViewModel: NSObject {
...
+            
+            let pendingTransactions = allPendingTransactions.filter {
+                $0.history.change?.asset.symbol.caseInsensitiveCompare(self.token.symbol) == .orderedSame &&
+                $0.address == maskUserDefaults.defaultAccountAddress &&
+                $0.networkId == maskUserDefaults.network.networkId
+            }.compactMap {
+                return $0.history
+            }
+            return (pendingTransactions + displayTransactions).sorted{ $0.timeAt > $1.timeAt }
+        }.receive(on: RunLoop.main)
```
Comment (jk234ert) [inline Maskbook/Scene/Wallet/TokenDetail/ViewModel/TokenDetailViewModel.swift]:
```suggestion
        }
        .receive(on: RunLoop.main)
```

File: Maskbook/Scene/Wallet/WalletSend/SendTransactionCofirmPopViewController.swift [hunk 4]
```diff
@@ -311,8 +314,15 @@ final class SendTransactionCofirmPopViewController: UIViewController {
                      self.gasFeeEthLabel.text = "\(costEth) \(maskUserDefaults.network.mainToken?.symbol ?? "")"
                      let tokenValue = NSDecimalNumber(string: self.viewModel.selectedTokenPublisher.value?.calculateAmountValue(quantity: self.amount))
                      let gasValue = NSDecimalNumber(string: EthUtil.getGasFee(gwei: gasPriceValue, gasLimit: gasLimitValue, price: token?.price as! Double))
+                     
                      let roundBehavior = NSDecimalNumberHandler(roundingMode: .plain, scale: 2, raiseOnExactness: false, raiseOnOverflow: false, raiseOnUnderflow: false, raiseOnDivideByZero: false)
-                     self.totalPriceLabel.text = "\(maskUserDefaults.currency.symbol)\(tokenValue.adding(gasValue).rounding(accordingToBehavior: roundBehavior).stringValue)"
+                     let totalValue = tokenValue.adding(gasValue).rounding(accordingToBehavior: roundBehavior)
+                     if totalValue.compare(NSDecimalNumber(string: "0.01")) == .orderedAscending{
```
Comment (jk234ert) [inline Maskbook/Scene/Wallet/WalletSend/SendTransactionCofirmPopViewController.swift:320]: Could we use `token.displayBalance` instead here?

File: Maskbook/Scene/Wallet/WalletSend/SendTransactionCofirmPopViewController.swift [hunk 5]
```diff
@@ -332,15 +342,48 @@ final class SendTransactionCofirmPopViewController: UIViewController {
     private func handleConfirmClicked(password: String) {
                 guard let fromAddress = maskUserDefaults.defaultAccountAddress  else { print("fromAddress is illegal"); return }
                 guard let toAddress = self.toAddress else { print("address is empty"); return }
+                guard let token = self.viewModel.selectedTokenPublisher.value else { return }
+
 
                 self.sendStatus = .sending
-                self.viewModel.sendTransaction(password: password, amount: self.amount, toAddress: toAddress, fromAddress: fromAddress) { [weak self] result in
+                
+        if (self.nonce != nil) {
```
Comment (jk234ert) [inline Maskbook/Scene/Wallet/WalletSend/SendTransactionCofirmPopViewController.swift]:
```suggestion
        if let nonce = nonce {
```

File: Maskbook/Scene/Wallet/WalletSend/SendTransactionConfirmViewController.swift [hunk 6]
```diff
@@ -473,8 +473,11 @@ class SendTransactionConfirmViewController: BaseViewController {
                 guard let gasPrice = gasPrice else { return }
                 guard let price = gasFeeToken?.price else { return }
                 let gasFee = EthUtil.getGasFee(gwei: gasPrice, gasLimit: gasLimit, price: price.doubleValue)
-                self?.feeButton.setTitle("\(L10n.Scene.Sendtransaction.Gasprice.fee) :\(maskUserDefaults.currency.symbol)\(gasFee)", for: .normal)
-                
+                if NSDecimalNumber(string: gasFee).compare(NSDecimalNumber(string: "0.01")) == .orderedAscending {
```
Comment (jk234ert) [inline Maskbook/Scene/Wallet/WalletSend/SendTransactionConfirmViewController.swift:476]: `Token.displayBalance`?

File: Maskbook/Wallet/WalletService/PendTransacation/PendTransactionManager.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import BigInt
+
+class PendTransactionManager {
+
+    private var pendList:[PendTransactionModel] = []
+    private var timer: Timer?
+    var subscriptions: Set<AnyCancellable> = []
+    public static let shared = PendTransactionManager()
+    
+    public var pendTransactions: CurrentValueSubject<[PendTransactionModel], Never> = CurrentValueSubject([])
```
Comment (jk234ert) [inline Maskbook/Wallet/WalletService/PendTransacation/PendTransactionManager.swift]: 1. `pendList` and `pendTransactions` could be 1 `CurrentValueSubject` as the unified data source IMO.
2. `subscriptions` not used?

File: Maskbook/Wallet/WalletService/PendTransacation/PendTransactionManager.swift [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+                    if self.pendList.isEmpty {
+                        self.timer?.invalidate()
+                    }
+                    
+                    for (index, pendingTranscation) in self.pendList.enumerated() {
+                        web3Provier.getTransactionReceiptPromise(pendingTranscation.txHash).done { transactionReceipt in
+                            if transactionReceipt.status != .notYetProcessed {
+                                self.pendList.remove(at: index)
+                            }
+                            self.pendTransactions.send(self.pendList)
```
Comment (jk234ert) [inline Maskbook/Wallet/WalletService/PendTransacation/PendTransactionManager.swift]: It's better if we could have a `PassThroughSubject` that sends the status change events of pending TXs.
Like:
```
var pendingTxFinishEvents = PassthroughSubject<PendTransactionModel, ConfimState>()
```
Once we found a pending tx finishes (either confirmed or failed), we:
1. modify the pendingTxList.
2. We emit the event with the tx and the latest status.

File: Maskbook/Wallet/WalletService/PendTransacation/PendTransactionManager.swift [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                        }
+                    }
+                }
+            }
+    }
+    
+    private func stop() {
+        timer?.invalidate()
+    }
```
Comment (jk234ert) [inline Maskbook/Wallet/WalletService/PendTransacation/PendTransactionManager.swift:59]: `stop` not used?

**Review by jk234ert — CHANGES_REQUESTED** [review]

File: Maskbook/Scene/Wallet/WalletSend/SendTransactionCofirmPopViewController.swift (hunk 4 above)
Comment (hydraCy) [reply Maskbook/Scene/Wallet/WalletSend/SendTransactionCofirmPopViewController.swift:320]: Here is gas + token.balance

**Review by jk234ert — APPROVED** [review]

**Pull request merged**